From 77d6953f0d4aa631c8f4b83f07141c26dd97ae4b Mon Sep 17 00:00:00 2001
From: Lin Jinhan <troy.lin@rock-chips.com>
Date: Fri, 10 Aug 2018 16:22:03 +0800
Subject: [PATCH] media: soc_camera: ov9750 support as-master inexistence

as-master = <1>: will set ov9750 as master mode
as-master = <0>: will set ov9750 as slave mode
as-master inexistence: will set ov9750 as normal mode

Change-Id: Ibea5769be6c126b9365ed9333ac931d20ca20f2f
Signed-off-by: Lin Jinhan <troy.lin@rock-chips.com>
---
 .../i2c/soc_camera/rockchip/ov9750_v4l2-i2c-subdev.c  | 11 +++++++----
 .../media/i2c/soc_camera/rockchip/ov_camera_module.h  |  2 +-
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/media/i2c/soc_camera/rockchip/ov9750_v4l2-i2c-subdev.c b/drivers/media/i2c/soc_camera/rockchip/ov9750_v4l2-i2c-subdev.c
index 69b4be96ce66..3e7476860fd2 100644
--- a/drivers/media/i2c/soc_camera/rockchip/ov9750_v4l2-i2c-subdev.c
+++ b/drivers/media/i2c/soc_camera/rockchip/ov9750_v4l2-i2c-subdev.c
@@ -341,7 +341,7 @@ static struct ov_camera_module_config ov9750_configs[] = {
 
 static int ov9750_dual_mode(struct ov_camera_module *cam_mod)
 {
-	if (cam_mod->as_master) {
+	if (cam_mod->as_master == 1) {
 		ov_camera_module_write_reg(cam_mod, 0x3002, 0xa1);
 		ov_camera_module_write_reg(cam_mod, 0x3007, 0x02);
 		ov_camera_module_write_reg(cam_mod, 0x3816, 0x00);
@@ -350,10 +350,12 @@ static int ov9750_dual_mode(struct ov_camera_module *cam_mod)
 		ov_camera_module_write_reg(cam_mod, 0x3819, 0x01);
 		ov_camera_module_write_reg(cam_mod, 0x3823, 0x00);
 		ov_camera_module_write_reg(cam_mod, 0x3824, 0x00);
-	} else {
+	} else if (cam_mod->as_master == 0) {
 		ov_camera_module_write_reg(cam_mod, 0x3002, 0x21);
 		ov_camera_module_write_reg(cam_mod, 0x3823, 0x48);
 		ov_camera_module_write_reg(cam_mod, 0x3824, 0x11);
+	} else {
+		;/* do nothing */
 	}
 	return 0;
 }
@@ -971,6 +973,7 @@ static int ov9750_probe(struct i2c_client *client,
 	const struct i2c_device_id *id)
 {
 	int ret = 0;
+	int as_master = -1;
 	struct device_node *np = of_node_get(client->dev.of_node);
 
 	dev_info(&client->dev, "probing cam_num:%d 0x%x\n",
@@ -982,8 +985,8 @@ static int ov9750_probe(struct i2c_client *client,
 
 	ov9750[cam_num].custom = ov9750_custom_config;
 
-	of_property_read_u32(np, "as-master", &ret);
-	ov9750[cam_num].as_master = (ret == 0) ? false : true;
+	ret = of_property_read_u32(np, "as-master", &as_master);
+	ov9750[cam_num].as_master = (ret == 0) ? as_master : -1;
 	cam_num++;
 
 	dev_info(&client->dev, "probing successful\n");
diff --git a/drivers/media/i2c/soc_camera/rockchip/ov_camera_module.h b/drivers/media/i2c/soc_camera/rockchip/ov_camera_module.h
index 11644dc47834..1ae6576a1b74 100644
--- a/drivers/media/i2c/soc_camera/rockchip/ov_camera_module.h
+++ b/drivers/media/i2c/soc_camera/rockchip/ov_camera_module.h
@@ -206,7 +206,7 @@ struct ov_camera_module {
 	u32 rotation;
 	void *pltfm_data;
 	bool inited;
-	bool as_master;
+	int as_master;
 	struct mutex lock;
 };
 
-- 
2.35.3

